Move trends plots inside of "Custom" section

julieg18 opened the pull request:
# 1/3 main <= this <= #3466 <= #3491
## Demo

https://user-images.githubusercontent.com/43496356/224517013-fbc554e8-8c81-43c6-91b2-d8b5412e960c.mov

Part of #3373

julieg18 commented on extension/src/plots/model/collect.ts:
```diff
@@ -282,69 +295,20 @@ const collectCustomPlotData = (
+): CustomPlot[] => {
+  return plotsOrderValues
+    .map((plotOrderValue): CustomPlot => {
+      if (isCheckpointValue(plotOrderValue)) {
```
Since CustomPlot has two types, we end up needing to this kind of check multiple times throughout  code, but is there a better way to handle the different types?

sroy3 replied:
Maybe adding a boolean value to `plotOrderValue` to easily check. Do we actually have to keep them separated throughout the code if the only difference is that one has a `param` property and not the other?

julieg18 commented on extension/src/plots/model/index.ts:
```diff
@@ -498,26 +457,36 @@ export class PlotsModel extends ModelWithPersistence {
+    if (!colors) {
+      return plots
+        .filter(plot => !isCheckpointPlot(plot))
+        .map(
```
This function can definitely be simplified and only use one `map` instead of two.

julieg18 commented on extension/resources/walkthrough/plots.md:
```diff
@@ -66,25 +66,28 @@ templates], which may be predefined (e.g. confusion matrix, linear) or custom
+can choose between two types of plots, "Checkpoint Trend" and "Metric Vs Param".
+"Checkpoint Trend" plots compare a chosen [metric] value per epoch if
+[checkpoints] are enabled and "Metric Vs Param" plots compare a chosen metric
+and param across experiments.
```
Any ideas for improving text or plot titles?

julieg18 commented on extension/src/plots/model/collect.test.ts:
```diff
@@ -3,25 +3,31 @@ import omit from 'lodash.omit'
 import isEmpty from 'lodash.isempty'
```
> The test coverage on the diff in this pull request is 91.2% (85% is the threshold).

I looked over the code and everything looked pretty covered. Any thing I could have missed to improve the coverage?

julieg18 commented on webview/src/plots/components/App.test.tsx:
```diff
@@ -515,21 +471,22 @@ describe('App', () => {
     renderAppWithOptionalData({
-      checkpoint: checkpointPlotsFixture
+      comparison: comparisonTableFixture,
+      custom: customPlotsFixture
```
`custom` won't render currently without additional data. Planning to changing that and will adjust tests according.

mattseddon reviewed (commented):
Made a first pass, took me a couple of hours. Will need to have another go at it. A lot of the bigger items can be done as follow-ups.

mattseddon commented:
@julieg18 @sroy3 

Moving between having a custom checkpoint trend plot in the custom section on this branch and `main` caused plots to be completely broken for me until I ran `Reset Persisted State and Reload Window`... beware, I thought this was a P0.

julieg18 commented on extension/src/plots/model/index.ts:
```diff
@@ -95,27 +96,11 @@ export class PlotsModel extends ModelWithPersistence {
       DEFAULT_SECTION_COLLAPSED
     )
     this.comparisonOrder = this.revive(PersistenceKey.PLOT_COMPARISON_ORDER, [])
-    this.selectedMetrics = this.revive(
```
Yes, there currently isn't a way to hide custom plots. I thought it was agreed that we wanted to take off the menu logic but maybe I misunderstood something 🤔

julieg18 commented on extension/src/plots/model/index.ts:
```diff
@@ -173,67 +157,64 @@ export class PlotsModel extends ModelWithPersistence {
+        ({
+          ...rest,
+          // type is possibly undefined if state holds an older version of custom plots
+          type: type || CustomPlotType.METRIC_VS_PARAM
```
> Can this update be moved into update so that we can rid of it eventually?

Good idea!

julieg18 commented on extension/src/plots/model/custom.ts:
```diff
@@ -0,0 +1,21 @@
+  param: string
+}
+
+export type CustomPlotsOrderValue = CheckpointValue | MetricVsParamValue
```
> [Q] Would giving CheckpointValue a param (the same for all plots) simplify a lot of the code around switching based around isCheckpointValue?

Good idea!

julieg18 commented on extension/src/plots/webview/messages.ts:
```diff
@@ -121,9 +117,37 @@ export class WebviewMessages {
-  private setSelectedMetrics(metrics: string[]) {
-    this.plots.setSelectedMetrics(metrics)
-    this.sendCheckpointPlotsAndEvent(EventName.VIEWS_PLOTS_METRICS_SELECTED)
+  public async addMetricVsParamPlot(): Promise<
```
I think that's a typo. Will correct it!

julieg18 commented on extension/src/plots/webview/messages.ts:
```diff
@@ -121,9 +117,37 @@ export class WebviewMessages {
+    })
+
+    if (plotAlreadyExists) {
+      return Toast.showError('Custom plot already exists.')
```
Yes and it's already a task in #3373

julieg18 commented on webview/src/plots/components/App.test.tsx:
```diff
@@ -72,18 +71,16 @@ jest.mock('../../shared/components/dragDrop/currentTarget', () => {
-  createSpec: () => ({
+jest.mock('./customPlots/util', () => ({
+  createCheckpointSpec: () => ({
     $schema: 'https://vega.github.io/schema/vega-lite/v5.json',
```
I believe I got an error when I tried to make it a constant.

sroy3 commented on extension/src/plots/model/collect.ts:
```diff
@@ -235,33 +238,43 @@ export const collectCheckpointPlotsData = (
+      type: CustomPlotType.METRIC_VS_PARAM
+    }),
     metric: metric.slice(ColumnType.METRICS.length + 1),
     param: param.slice(ColumnType.PARAMS.length + 1),
```
Why/what are we slicing here? I can probably follow the code to know why, but a comment would help. Why aren't we doing it for the id? Can that be a problem later on?

julieg18 replied:
> Why/what are we slicing here?

We're taking off the `params:` and `metrics:` from the paths.

>  I can probably follow the code to know why, but a comment would help. Why aren't we doing it for the id? Can that be a problem later on?

The id uses these path prefixes since the `customPlotsOrder` uses them. But you're right, the two different paths are confusing. We can simplify things by having `customPlotsOrder` and `customPlotId` stop using the prefixes.

sroy3 commented on extension/src/plots/webview/messages.ts:
```diff
@@ -244,19 +275,34 @@ export class WebviewMessages {
+      customPlots,
+      'id'
+    ).map(plot =>
+      isCheckpointValue(plot)
```
The check seems useless here. `metric` is the same, `type` already exists as you're checking for it here, `param` can be set to undefined depending on the `type`

julieg18 replied:
This part is simplified now by adding a param to `Checkpoint` type plot values. I did look into making the param optional but it seemed to cause extra code complexity elsewhere.

sroy3 commented on webview/src/plots/components/customPlots/customPlotsSlice.ts:
```diff
@@ -43,13 +44,14 @@ export const customPlotsSlice = createSlice({
       return {
         ...state,
         ...statePayload,
+        colors: colors || { domain: [], range: [] },
```
`{ domain: [], range: [] }` could be a const. `const initialColorsState = { domain: [], range: [] }`.

sroy3 commented on webview/src/plots/hooks/useGetPlot.ts:
```diff
@@ -10,24 +10,15 @@ import { PlainObject, VisualizationSpec } from 'react-vega'
 ) => {
-  const isPlotWithSpec =
-    section === Section.CHECKPOINT_PLOTS || section === Section.CUSTOM_PLOTS
+  const isPlotWithSpec = section === Section.CUSTOM_PLOTS
```
This used to be higher and used for the check on line 13 as well. I think most of this file can be reverted to what it used to before duplication of checkpoints and custom (https://github.com/iterative/vscode-dvc/commit/165108829bdfaf35d5c8df39fc5efdcfa49c9772#diff-164992d57e219d2123695397d166b9b166a6aff3833b781b4241e22b5078e0d7)

sroy3 commented:
Not sure if it's just because my demo project was not properly set up when I opened the branch, but I can only see custom plots when there is an experiment running and all plots look the same.


https://user-images.githubusercontent.com/3683420/224716434-381248ff-d1ac-439f-9b46-76721911d7a5.mov

julieg18 commented on extension/src/plots/webview/messages.ts:
```diff
@@ -180,38 +204,50 @@ export class WebviewMessages {
+    if (plotType === CustomPlotType.CHECKPOINT) {
+      void this.addCheckpointPlot()
+    } else {
+      void this.addMetricVsParamPlot()
```
Updated!

julieg18 commented on extension/src/plots/model/quickPick.ts:
```diff
@@ -68,3 +116,21 @@ export const pickMetricAndParam = async (columns: Column[]) => {
+  })
+
+  if (!metric) {
+    return
```
I believe it's covered in [`quickPick.test.ts`](https://github.com/iterative/vscode-dvc/pull/3404/files#diff-3828a8ed605862eb03ab7072a8e3a314bff9fe15758544850d94523adea3c57dR223)

mattseddon replied:
I'm just going off what code coverage is telling me:

<img width="660" alt="image" src="https://user-images.githubusercontent.com/37993418/224799790-07f73bb3-1f1f-460c-abc2-568d0aac2220.png">

julieg18 commented on extension/resources/walkthrough/live-plots.md:
```diff
@@ -33,8 +33,6 @@ for epoch in range(NUM_EPOCHS):
 language and any tool.
 
-💡 `Trends` section of the plots dashboard is being updated automatically based
+💡 `Custom` section of the plots dashboard is being updated automatically based
```
Adjusted the text! Hopefully, it's more clear now :)

julieg18 commented:
> Not sure if it's just because my demo project was not properly set up when I opened the branch, but I can only see custom plots when there is an experiment running and all plots look the same.

@sroy3, I'm a little confused by your statement since looking at your demo, I see the custom plots the entire time... What am I misunderstanding? 

I do notice that the trends plots don't seem to be live updating but I think that is due to the running experiment not getting selected to the very end... Looks like that happens on main as well on my end.

sroy3 commented:
> > Not sure if it's just because my demo project was not properly set up when I opened the branch, but I can only see custom plots when there is an experiment running and all plots look the same.
> 
> @sroy3, I'm a little confused by your statement since looking at your demo, I see the custom plots the entire time... What am I misunderstanding?
> 

They are there, but they are empty. No matter what metric or param I choose, all plots are the same inside.

julieg18 commented:
Ok, thanks for the reviews everyone! I've resolved most of the comments. 

What's left:

- [ ] Delete old Persistance metrics state (https://github.com/iterative/vscode-dvc/pull/3404#discussion_r1133350817)
- [ ] Adjust custom plots order state (https://github.com/iterative/vscode-dvc/pull/3404#discussion_r1133350999)
- [ ] Investigate test (https://github.com/iterative/vscode-dvc/pull/3404#discussion_r1133370484)
- [ ] Simplify metric and param path logic (https://github.com/iterative/vscode-dvc/pull/3404#discussion_r1133905101)

mattseddon commented on extension/src/plots/model/index.ts:
```diff
@@ -498,26 +458,21 @@ export class PlotsModel extends ModelWithPersistence {
+      filteredPlots = plots.filter(plot => !isCheckpointPlot(plot))
+    }
+    return filteredPlots.map(plot =>
+      collectCustomPlotData(
```
[C] What I meant here was that you don't have to filter and then map you could do it in a single loop:

```
  private getCustomPlotsData(
    plots: CustomPlot[],
    colors: ColorScale | undefined
  ): CustomPlotData[] {
    const selectedExperimentsExist = !!colors
    const filteredPlots: CustomPlotData[] = []
    for (const plot of plots) {
      if (!selectedExperimentsExist && isCheckpointPlot(plot)) {
        continue
      }
      filteredPlots.push(
        collectCustomPlotData(
          plot,
          colors,
          this.getNbItemsPerRow(Section.CUSTOM_PLOTS)
        )
      )
    }
    return filteredPlots
  }
```

```

julieg18 replied:
Fixed, thanks for the clarification!

mattseddon commented on extension/src/plots/model/custom.ts:
```diff
@@ -0,0 +1,25 @@
+  param: string
+}
+
+export const isCheckpointValue = (
```
[I] Most of the usage of this can be replaced with `isCheckpointPlot`. E.g 

```
  const plotsItems = plots.map(plot =>
    isCheckpointValue(plot.type)
      ? getCheckpointPlotItem(plot.metric)
      : getMetricVsParamPlotItem(plot.metric, plot.param)
  )
```

```
      param: isCheckpointValue(plot.type)
        ? plot.param
        : buildMetricOrParamPath(ColumnType.PARAMS, plot.param),
```

julieg18 replied:
`isCheckpointPlot` wouldn't work for those instances since they're inspecting a order item, not a `CheckpointPlot` 🤔

mattseddon replied:
😕. Ok, the problem is that `plots` are not `plots` they `CustomPlotsOrderValue`s. The variable name should reflect what they are.

<img width="559" alt="image" src="https://user-images.githubusercontent.com/37993418/224859359-160be343-f097-4b06-8d8c-85b30197d843.png">

julieg18 replied:
Oh, yes, you're right! I need to take a break, my brain is not thinking straight 😅

mattseddon commented on webview/src/plots/components/customPlots/CustomPlot.tsx:
```diff
@@ -13,26 +18,38 @@ interface CustomPlotProps {
+  if (isCheckpointPlot(plot)) {
+    return createCheckpointSpec(plot.yTitle, plot.param, colors)
+  }
+  return createMetricVsParamSpec(plot.yTitle, plot.param)
```
[Q] Does this still need to be in the webview? Can it be moved back to the extension?

The reason that I ask is that we want to limit the number of places `isCheckpointPlot`/`isCheckpointValue` get called. Having the logic spread out through different parts of the codebase will lead to bugs. 

Note: Not everyone agrees but type-checking can be considered a code smell. I think it is ok if it is contained to a single part of the codebase/a few files but the more it spreads the more likely it is to cause bugs.

julieg18 replied:
> [Q] Does this still need to be in the webview? Can it be moved back to the extension?

It looks like it can be sent via the extension... But I'm unaware of why it was in the webview in the first place 🤔

julieg18 commented:
> They are there, but they are empty. No matter what metric or param I choose, all plots are the same inside.

Weird, things look correct on my end: 

https://user-images.githubusercontent.com/43496356/224848463-0cf9f71f-2888-4527-880b-dc62e8e609cb.mov

@mattseddon, are the plots rendering correctly for you?

mattseddon reviewed (approved):
> Ok, thanks for the reviews everyone! I've resolved most of the comments.
> 
> What's left:
> 
> * [ ]  Delete old Persistance metrics state ([Move trends plots inside of "Custom" section #3404 (comment)](https://github.com/iterative/vscode-dvc/pull/3404#discussion_r1133350817))
> * [ ]  Adjust custom plots order state ([Move trends plots inside of "Custom" section #3404 (comment)](https://github.com/iterative/vscode-dvc/pull/3404#discussion_r1133350999))
> * [ ]  Investigate test ([Move trends plots inside of "Custom" section #3404 (comment)](https://github.com/iterative/vscode-dvc/pull/3404#discussion_r1133370484))
> * [ ]  Simplify metric and param path logic ([Move trends plots inside of "Custom" section #3404 (comment)](https://github.com/iterative/vscode-dvc/pull/3404#discussion_r1133905101))

- [ ]  Simplify metric and param path logic ([Move trends plots inside of "Custom" section #3404 (comment)](https://github.com/iterative/vscode-dvc/pull/3404#discussion_r1133905101))
- [ ] Duplicate code removed (https://github.com/iterative/vscode-dvc/pull/3404/files#r1133369194)

I am ok to merge this if the checkboxes directly above get addressed. The checkboxes below need to be added to the issue and the issue should remain open/get worked on until they are all finished.

- [ ] Delete old Persistance metrics state ([Move trends plots inside of "Custom" section #3404 (comment)](https://github.com/iterative/vscode-dvc/pull/3404#discussion_r1133350817))
- [ ]  Adjust custom plots order state ([Move trends plots inside of "Custom" section #3404 (comment)](https://github.com/iterative/vscode-dvc/pull/3404#discussion_r1133350999))
- [ ]  Investigate test ([Move trends plots inside of "Custom" section #3404 (comment)](https://github.com/iterative/vscode-dvc/pull/3404#discussion_r1133370484))
- [ ] Update walkthrough images after other PRs are merged (https://github.com/iterative/vscode-dvc/pull/3404/files#r1133343393)
- [ ] Check if spec creation can be moved to extension
- [ ] Consolidate type checking to smallest space/amount of functions
- [ ] Add code to add 1 trend and 1 metric vs param custom plot to e2e test (https://github.com/iterative/vscode-dvc/pull/3404/files#r1133364032)

mattseddon commented:
> > They are there, but they are empty. No matter what metric or param I choose, all plots are the same inside.
> 
> Weird, things look correct on my end:
> 
> @mattseddon, are the plots rendering correctly for you?

From the original screenshot:

<img width="1323" alt="image" src="https://user-images.githubusercontent.com/37993418/224850883-72ad0a60-73f0-4531-894d-a1c6fee3cbea.png">

They aren't empty. I think they just contain the same data for each of the experiments. Try updating params before running experiments?

Changing LR:

<img width="1728" alt="image" src="https://user-images.githubusercontent.com/37993418/224851347-92afa7f7-a49e-4103-9922-e0920407eb4f.png">

julieg18 commented:
> I am ok to merge this if the checkboxes directly above get addressed before the merge. The checkboxes below need to be added to the issue and the issue should remain open/get worked on until they are all finished.

Resolved the checkboxes above (one ended up getting merged in a secondary pr). I'll look over the code one final time tomorrow morning and, if there are no objections, merge this pr.

julieg18 commented on extension/src/plots/model/quickPick.ts:
```diff
@@ -1,40 +1,81 @@
+import { CustomPlotType } from '../webview/contract'
+
+const getMetricVsParamPlotItem = (metric: string, param: string) => {
+  const splitMetric = splitColumnPath(metric)
```
This needs to a `TYPE:` prefix.

sroy3 commented on webview/src/plots/components/customPlots/util.ts:
```diff
@@ -1,6 +1,99 @@
+            { field: 'group', title: 'name' },
+            {
+              field: 'y',
+              title: title.slice(Math.max(0, title.indexOf(':') + 1)),
```
Shouldn't this be done before `truncateVerticalTitle` (on the extension side)?

julieg18 replied:
Sounds like a good idea! Will add this to our followup checklist :)

sroy3 replied:
I feel like breaking changes introduces in this PR should be fixed in this PR, unless it's something really big or introduces a new concern.

julieg18 replied:
> I feel like breaking changes introduces in this PR should be fixed in this PR, unless it's something really big or introduces a new concern.

Of course! I technically didn't add this line of code so I didn't realize it was a breaking change :)

julieg18 commented on extension/src/plots/model/collect.ts:
```diff
@@ -235,38 +247,47 @@ export const collectCheckpointPlotsData = (
-  `custom-${metric}-${param}`
-
-const collectCustomPlotData = (
+const collectMetricVsParamPlot = (
```
Apologies, I'm a little confused by your question. Both of those functions do very different things at the moment (one relies on output data while other relies on experiment data)... Though we could definitely look into consolidating how we get the custom plot data. I think we could gather checkpoint data with experiment data possibly 🤔

sroy3 replied:
I mean I don't understand why every time there is something done to custom plots it has to be separated between `checkpoint` and `metricVsParam` when most of the time the difference between could be handle by the same function the type being one or the other. We don't actually have to make the distinction between both in most places.

sroy3 replied:
I would rather not introduce a temporary data structure to be replaced later. This usually creates tech debt and here the separation introduces complexity that we'll have to deal with until it is fixed.

julieg18 replied:
When you say temporary data structure, are you referring to how the checkpoints plots are being held in an object?

julieg18 replied:
> I would rather not introduce a temporary data structure to be replaced later. This usually creates tech debt and here the separation introduces complexity that we'll have to deal with until it is fixed.

I understand not wanting to create a tech debt, but I really think we need to get this merged. But, the longer we wait the harder it's going to be to merge it because it alters so many different plot files. What do you think?

sroy3 replied:
I get it, it works as expected when we run the extension and there will be a lot of conflicts, but we shouldn't rush to merge things because of this unless it fixes a bug. Most of the conflicts you'll encounter are going to be because we created a temporary duplication before (custom and checkpoints). Creating temporary data structures will simply make more merge conflicts later. This PR is about moving trend plots inside custom plots, why keep treating them separately then? In other words, I feel like this is the heart of the PR. I see it more as a fix inside this PR concern than a follow up. If everyone else is fine merging this, then I'll let go.

julieg18 replied:
Makes sense... @mattseddon what do you think?

If Matt agrees, I'll go ahead and convert this into a draft and work on a second pr that has both checkpoint/custom data gathered with experiment data. That should take care of the temporary data structure (`customCheckpoints`) we have.

mattseddon replied:
I think we can delay merging in order to get this right.

A few takeaways for me and to discuss in the retro:

1. We probably should not have merged https://github.com/iterative/vscode-dvc/pull/3342.
2. Best way forwards after https://github.com/iterative/vscode-dvc/pull/3342 may have been to chain off a revert.
3. When PRs are this big we need to shorten the feedback loop (pair/mob is best option).
4. We need to fight against making so many changes in a single PR. Find some other way to discuss data structures/approaches early and avoid thrashing.

🙏🏻

julieg18 commented:
https://github.com/iterative/vscode-dvc/pull/3404#discussion_r1135538551 has been resolved with #3466 and #3491 (both approved and ready to be merged into this pr)! 

Here's a (hopefully complete) list of what's left:

- [ ] Delete old Persistance metrics state (https://github.com/iterative/vscode-dvc/pull/3404#discussion_r1133350817)
- [ ] Adjust custom plots order state (https://github.com/iterative/vscode-dvc/pull/3404#discussion_r1133350999)
- [ ] Investigate test (https://github.com/iterative/vscode-dvc/pull/3404#discussion_r1133370484)
- [ ] Update walkthrough images after other PRs are merged (https://github.com/iterative/vscode-dvc/pull/3404/files#r1133343393)
- [ ] Check if spec creation can be moved to extension (https://github.com/iterative/vscode-dvc/pull/3404#discussion_r1134686640)
- [ ] Consolidate type checking to the smallest space/amount of functions (https://github.com/iterative/vscode-dvc/pull/3404#discussion_r1134686640) (Most of this has been done already but more to do)
- [ ] Add code to add 1 trend and 1 metric vs param custom plot to e2e test (https://github.com/iterative/vscode-dvc/pull/3404/files#r1133364032)

Anything else that should get taken care of before we merge this? Or did I miss anything when I took care of merge conflicts?

julieg18 commented on webview/src/plots/components/App.test.tsx:
```diff
@@ -298,12 +278,11 @@ describe('App', () => {
     const loading = await screen.findAllByText('Loading...')
 
-    expect(loading).toHaveLength(3)
+    expect(loading).toHaveLength(2)
```
The custom section does not have a loading state. Technically only trend type plots rely on revisions, but I can go ahead and add one. If there is a better way to handle loading state for `Custom`, we can discuss further in the followup pr!

sroy3 reviewed (approved):
I think you used to have something that blocked the following from happening before:


https://user-images.githubusercontent.com/3683420/226343127-090754fe-d10f-4317-9b7e-8db6c1d40df8.mov

Could be easy to just show the available choice from the list. (Can be a follow-up)

julieg18 commented:
> I think you used to have something that blocked the following from happening before:
> Could be easy to just show the available choice from the list. (Can be a follow-up)

Thanks for spotting that! Simple bug in the quick pick logic/tests that I missed earlier. Will fix it before merging since I'd prefer to keep users' custom plot state to not have copies of the same plot in it :) Will keep current logic for now and save choosing available choices for a followup.

julieg18 commented:
Code climate seems to be failing with 

```
ERROR: Coverage for lines (94.9%) does not meet global threshold (95%)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
ERROR: "cover:report" exited with 1.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Error: The process '/usr/bin/xvfb-run' failed with exit code 1
/home/runner/work/_actions/paambaati/codeclimate-action/v3.2.0/node_modules/@actions/exec/lib/toolrunner.js:592
                error = new Error(`The process '${this.toolPath}' failed with exit code ${this.processExitCode}`);
                        ^

Error: The process '/usr/bin/xvfb-run' failed with exit code 1
    at ExecState._setResult (/home/runner/work/_actions/paambaati/codeclimate-action/v3.2.0/node_modules/@actions/exec/lib/toolrunner.js:592:25)
    at ExecState.CheckComplete (/home/runner/work/_actions/paambaati/codeclimate-action/v3.2.0/node_modules/@actions/exec/lib/toolrunner.js:575:18)
    at ChildProcess.<anonymous> (/home/runner/work/_actions/paambaati/codeclimate-action/v3.2.0/node_modules/@actions/exec/lib/toolrunner.js:469:27)
    at ChildProcess.emit (node:events:527:28)
    at maybeClose (node:internal/child_process:1092:16)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:302:5)
Error: 🚨 Coverage run failed!
```

First guess is the issue is amount of lines being covered:

```
ERROR: Coverage for lines (94.9%) does not meet global threshold (95%)
```

Will see if upping the code coverage fixes the issue!

julieg18 commented:
Updated the tests based of code reviews and looking over things myself:
* fixed `pickMetric` not being 100% covered
* added tests for `addCustomPlot` and `removeCustomPlot` ending early due to user not selecting things
* added additional tests for functions in `custom.test.ts` that appeared to be missing coverage


But code coverage is still failing:

```
ERROR: Coverage for lines (94.96%) does not meet global threshold (95%)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
ERROR: "cover:report" exited with 1.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
/home/runner/work/_actions/paambaati/codeclimate-action/v3.2.0/node_modules/@actions/exec/lib/toolrunner.js:592
Error: The process '/usr/bin/xvfb-run' failed with exit code 1
                error = new Error(`The process '${this.toolPath}' failed with exit code ${this.processExitCode}`);
                        ^
Error: The process '/usr/bin/xvfb-run' failed with exit code 1
    at ExecState._setResult (/home/runner/work/_actions/paambaati/codeclimate-action/v3.2.0/node_modules/@actions/exec/lib/toolrunner.js:592:25)
    at ExecState.CheckComplete (/home/runner/work/_actions/paambaati/codeclimate-action/v3.2.0/node_modules/@actions/exec/lib/toolrunner.js:575:18)
    at ChildProcess.<anonymous> (/home/runner/work/_actions/paambaati/codeclimate-action/v3.2.0/node_modules/@actions/exec/lib/toolrunner.js:469:27)
    at ChildProcess.emit (node:events:527:28)
    at maybeClose (node:internal/child_process:1092:16)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:302:5)
Error: 🚨 Coverage run failed!
```

Last things I'm seeing regarding testing is:

- [ ] Add tests for lines like this in collect.test.ts: `exp.name || exp.label`
- [ ] Add tests for when user recreates the same custom plot

I'll start working on the followup on stopping a user from recreating custom plots since I think it will take about the same time as just creating tests for our current solution. But is there something else that could be going on with CodeClimate that I'm missing? cc @sroy3, @mattseddon
